### PR TITLE
JitIL: Clean up memory address immediate offset optimisation.

### PR DESCRIPTION
**Pull request by phire**
This is an alternative to #2272, trying to attack the root cause of the bug instead of just correcting it.

Instead digging into the IR tree during codegen to extract base/offset address operands out of an add instruction, we extract the operands during folding and store the offset as an extra operand in load/store instructions. This required implementing 3 operand instructions for stores.

As an added benefit, this allowed a massive simplification of the Mark/BuildMemAddress functions.

I also removed an mov instruction in some memory loads.

### CODE REVIEW

File: Source/Core/Core/PowerPC/JitILCommon/IR.cpp [hunk 1]
```diff
@@ -1199,6 +1203,20 @@ InstLoc IRBuilder::FoldFallBackToInterpreter(InstLoc Op1, InstLoc Op2)
 	return EmitBiOp(FallBackToInterpreter, Op1, Op2);
 }
 
+std::pair<InstLoc, InstLoc> IRBuilder::FoldMemoryAddress(InstLoc addr)
+{
+	if (getOpcode(*addr) == Add && isImm(*getOp2(addr))) {
```
Comment (FioraAeterna) [inline Source/Core/Core/PowerPC/JitILCommon/IR.cpp]: nitpick: the { is inconsistent here

File: Source/Core/Core/PowerPC/JitILCommon/IR.cpp [hunk 2]
```diff
@@ -1254,11 +1273,25 @@ InstLoc IRBuilder::FoldBiOp(unsigned Opcode, InstLoc Op1, InstLoc Op2, unsigned
...
 		case FDSub:
 			return FoldDoubleBiOp(Opcode, Op1, Op2);
+		case Store8:
+		case Store16:
+		case Store32:
+		case StoreSingle:
+		case StoreDouble:
+		{
+			auto mem = FoldMemoryAddress(Op2);
+			return EmitTriOp(Opcode, Op1, mem.first, mem.second);
```
Comment (magumagu) [inline Source/Core/Core/PowerPC/JitILCommon/IR.cpp]: This is sort of abusing the intent of FoldBiOp etc.: given that Store8 is a three-operand instruction, it should be going through FoldTriOp.  (The difference doesn't really matter for now, but the intent was that we could merge blocks together by just refolding the IR, without going back to the original PPC asm.)

File: Source/Core/Core/PowerPC/JitILCommon/IR.cpp [hunk 3]
```diff
@@ -1199,6 +1203,21 @@ InstLoc IRBuilder::FoldFallBackToInterpreter(InstLoc Op1, InstLoc Op2)
...
+	if (getOpcode(*addr) == Add && isImm(*getOp2(addr)))
+	{
+		// We have Base + Offset which can be used directly as a memory address.
+		// Forget about the add and use the operands directly
+		return std::make_pair(getOp1(addr), getOp2(addr));
+	}
+	else
+	{
+		// Otherwise we just use the result of the add with a zero offset
+		return std::make_pair(addr, EmitIntConst(0));
```
Comment (magumagu) [inline Source/Core/Core/PowerPC/JitILCommon/IR.cpp]: This probably emits a lot of "0" constants... not really an issue now, but maybe an issue if we start saving IR.

**Comment by magumagu** [comment]
I'm not confident this is the right approach... it covers the current x86 reg+imm addressing optimization, but it breaks down if we try to add support for more complicated addressing modes.  And it's not nearly as clean if we support JITIL for other architectures, which can't encode a general 32-bit immediate into a load instruction.

I think the current x86-specific code looks more complicated than it actually is because there isn't an abstraction for instruction selection.

I guess in a more general sense, the question is whether the current IR is at the right level of abstraction from machine-specific code; should we be introducing target-specific logic into IR folding?

File: Source/Core/Core/PowerPC/JitILCommon/IR.cpp (hunk 2 above)
Comment (phire) [inline Source/Core/Core/PowerPC/JitILCommon/IR.cpp]: Ah, That's what the intent was.

I was originally using FoldTriOp but it created a whole lot more unused Zero Constant nodes.

File: Source/Core/Core/PowerPC/JitILCommon/IR.cpp (hunk 3 above)
Comment (phire) [inline Source/Core/Core/PowerPC/JitILCommon/IR.cpp]: we could implement a constant pooling system so multiple memory address point at the same zero.

But that's probably better done as a generic common expression pass.

**Comment by phire** [comment]
The basic goal with this PR is to enforce that every instruction should be pointing at exactly the operands it uses, so register clearing logic can be made much more generic. 

Instead of storing a last used bit per operand for each instruction in IInfo, we would just use the last use the information in lastUsed, scanning over it after each instruction and clearing any register which isn't used any more.

**Comment by phire** [comment]
But there is the question about right level of abstraction. 

The way I envision this supporting other architectures which can't encode a 32-bit immediate into the load instruction is to deal with it at folding time, either enforcing that the offset must be zero (or in the range they actually support) or create separate instructions for different addressing modes and selecting them at folding time.

In this scheme, each IR instruction is self contained, it can't generate half an x86 instruction. If your arch supports a "multiply by 9, add a 7bit immediate then divide by 42" instruction you must create an IR instruction that matches and select it while folding.

These is a notable exception for immediate instructions.

Perhaps it would be better to implement a proper instruction section step.

**Comment by magumagu** [comment]
My guess would be that a proper instruction selection step would be the right direction: some sort of secondary representation will probably be necessary anyway to improve register allocation, and Fold() doesn't have the context to make decisions based on the number of uses of an expression.

On the other hand, bringing the result of Fold() closer to x86 reduces probably makes the JIT run faster if we aren't planning to add more complicated optimizations.

I'm not really planning to work on JITIL, so I'll leave it up to you.

**Comment by phire** [comment]
Closing this for now.

Going to look into doing proper instruction selection with a secondary IR.

**Pull request closed without merging**